### ISSUE TITLE
HTML Custom Elements createdCallback firing in wrong place

### ISSUE THREAD

**Issue by SergeBondar** [angular/angular#13699]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report
[ ] feature request
[ ] support request
```

**Current behavior**
I'm trying to integrate ChartIQ library which has HTML Custom Elements to my angular 2 application.
Some of these elements have logic in createdCallback, but when it is firing innerHTML of Custom element is empty. 

**Expected behavior**
If I use Custom Elements without Angular, createdCallback is firing when the full element (with all childrens added into it) inserted into DOM.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/OW500u9fr27aVteo3zJh

**What is the motivation / use case for changing the behavior?**
For now Angular 2 breaks some part of expected behavior of Custom Elements. It needs to be fixed to allow using Custom Elements in Angular apps.

**Please tell us about your environment:**
Windows 8, WebStorm, npm

* **Angular version:**
2.2.1

* **Browser:** 
Chrome 55
 
* **Language:** 
all

* **Node (for AoT issues):** `node --version` =   
6.7.0

**Comment by DzmitryShylovich** [comment]
As a workaround u can wrap it into `setTimeout`:
```
proto.createdCallback = function() {
      setTimeout(() => {
        console.log("innerHTML: " + this.innerHTML);
        console.log("attributes count: " + this.attributes.length);
      });
    };
```

**Comment by SergeBondar** [comment]
Unfortunately I have no accees to this functionality as I'm trying to use external library.

**Comment by DzmitryShylovich** [comment]
@tbosch you are working on a new view engine. do you think it's possible to address this issue?

**Comment by tbosch** [comment]
Is this working with React?

**Comment by tbosch** [comment]
(Just trying to understand possible solutions...)

**Comment by SergeBondar** [comment]
I haven't tried, but this article (https://cbateman.com/blog/web-components-in-angular-ember-and-react/) says that yes.

**Comment by DzmitryShylovich** [comment]
Still valid in v4 http://plnkr.co/edit/F7ztSukQ7HLoFVwQJuZ0?p=preview

**Comment by zoechi** [comment]
@SergeBondar a bit late, but just saw - if you need the attribute with binding to be added to the DOM you need to use attribute binding instead of property binding https://plnkr.co/edit/NMCLHQ4H1kVzDH1YbhNV?p=preview (`attr.value="{{value}}" attr.yes-label="{{yesLabel}}"`)

**Comment by tbosch** [comment]
In React, this is actually working, see https://plnkr.co/edit/32MPt6QDHM2xzGTJzfyn?p=info

So there are 2 cases:

1) a user placed content behind `*ngIf` / ... as children
2) a users placed static content as children (the case in the plunker)

1) can't be solved as the condition for `ngIf` might only become truthy at a later time or never at all
2) could be solved by separating dom creation from directive creation, as our directives rely on the DOM element they are on being attached already. However:
    - this would make our view creation slower (as we need 2 passes over all nodes)
    - this would be a breaking change as now all child elements are already created when a directive is created.

**Comment by chenthillrulz** [comment]
We are also facing an issue where all the property setter methods are getting called only after  attachedCallback in the webcomponent. It would be nice to receive all the intial set of properties before the attachedCallBack so that we can render the UI once. It becomes unnecessary to re-render the component for properties that are only set as part of component init.

Is this covered as part of this issue or should I need to raise a separate issue for the same ?

**Comment by robdodson** [comment]
In Custom Elements v1 the `createdCallback` has been replaced with the `constructor`. [According to the spec](https://www.w3.org/TR/custom-elements/#custom-element-conformance), you should not try to access the attributes or children of a Custom Element in the constructor. If your element accepts children you should give it a shadow root with a `<slot>` element, and [listen for the `slotchange` event](https://hayato.io/2016/shadowdomv1/#events-to-react-the-change-of-distributions) to know when those children are ready.

@chenthillrulz's issue sounds like a separate discussion but it also doesn't sound like angular is doing anything "wrong" in this case. I think it's kind of up to the framework when it decides to set properties.

**Comment by trotyl** [comment]
Superseded by https://github.com/angular/angular/issues/18909

**Comment by pkozlowski-opensource** [comment]
This is fixed in Angular v9 running ivy, see https://github.com/angular/angular/issues/18909#issuecomment-590868286
